### PR TITLE
Simplify load of J9Class flags off fix loadaddr

### PR DESCRIPTION
**Pull request by r30shah**
Changes in [1] optimizes the VFTLoad from an object if the object is of known fixed type. This transformation would convert such load from an object to load of constant address. This would change
how loading of isClassAndDepthFlags would look like from J9Class *. Consider following example,

```
n3745n    treetop
n3744n      new  jitNewObject[#91  helper Method] [flags 0x400 0x0 ]
n3743n        loadaddr  org/junit/internal/ExactComparisonCriteria[#748  Static]
n3802n    astore  <temp slot 2>[#758  Auto]
n3744n      ==>new
n3820n    NULLCHK on n3784n [#32]
n3788n      aloadi  <vft-symbol>[#307  Shadow]
n3784n        aload  <temp slot 2>[#758  Auto]
n3793n    ificmpne --> block_69 BBStart at n3799n ()
n3791n      iand (X>=0 cannotOverflow )
n3789n        iloadi  <isClassAndDepthFlags>[#304  Shadow 12]
n3788n          ==>aloadi
n3790n        iconst 0x40200000 (X!=0 X>=0 )
n3792n      iconst 0 (X==0 X>=0 X<=0 )
```

In above case, with the change to optimize VFTLoad into loadaddr, ValuePropagation first encounters `n3788n` and as it is loading J9Class * from the object of fixed type, it would convert it to
`loadaddr` node, loading a constant `J9Class *`. 
Before we added this transformation in Global Value Propagtion, we used to simplify loading of `isClassAndDepthFlags` field from J9Class * which and fold it to constant. Looking into the code [2], where we do this transformation, I see that it does it with the assumption that child of such indirect load would be VFTLoad from an object, and 
to see if it of known type or not and it will fold the load of such flags to constant if the VFT load is from object of known fixed type. This part of the code does not check if it is loading flags from
constant address which represents Class object of fixed type or not. Commit in this PR updates the simplifyJ9ClassFlags to check if the child of indirect load is load of constant of type class object.

[1]. https://github.com/eclipse/omr/commit/1818cb10df73ff42e54e21dbc0a2246e6e504ae2
[2]. https://github.com/eclipse/omr/blob/96a484d1b5b7c27f5061965a5651738996d3bc44/compiler/optimizer/VPHandlers.cpp#L1345-L1350

### CODE REVIEW

**Comment by r30shah** [comment]
@vijaysun-omr / @jdmpapin  Can I please get your review on these changes? 

For context, while looking into a failure on Z, I found out that due to change I made last year to fold VFT load of from object of fixed type to constant, We were missing out on simplifying the loading of `ClassFlags` from `J9Class *` as it expects the child to be VFT Load only. 
I am testing this changes through internal personal builds so marking it WIP till builds are finished.

File: compiler/optimizer/VPHandlers.cpp [hunk 1]
```diff
@@ -1342,19 +1342,26 @@ bool simplifyJ9ClassFlags(OMR::ValuePropagation *vp, TR::Node *node, bool isLong
...
-            {
-            cdfValue = vp->comp()->fej9()->getClassDepthAndFlagsValue(baseObject->getClassType()->getClass());
-            // if the type is a java.lang.Object, make the control branch to the call
-            //
-            if (baseObject->getClassType()->asFixedClass()->isJavaLangObject(vp))
-               cdfValue = TR::Compiler->cls.flagValueForFinalizerCheck(vp->comp());
-            }
+         baseNode = node->getFirstChild()->getFirstChild();
+         }
+      else if(node->getFirstChild()->getOpCode().isLoadAddr() && node->getFirstChild()->getSymbolReference()->getSymbol()->isClassObject())
```
Comment (jdmpapin) [inline compiler/optimizer/VPHandlers.cpp:1351]: I think this logic probably shouldn't have to inspect the structure of the child at all, not even with the original `if` just above. We should be able to just ask for the constraint of `node->getFirstChild()`, e.g.
```
TR::VPConstraint *base = vp->getConstraint(node->getFirstChild(), isGlobal);
if (base != NULL && base->isJ9ClassObject() == TR_yes && base->isFixedClass())
   ...
```

Comment (r30shah) [reply compiler/optimizer/VPHandlers.cpp:1351]: Yes, that should work as well. Thanks a lot for the suggestions. Let me try it out.

Comment (r30shah) [reply compiler/optimizer/VPHandlers.cpp]: @jdmpapin  Made changes in https://github.com/eclipse/omr/compare/f94a66015781949772cf21347d57a624b2af2bee..abdec2fc9577133c9b9f9c052b17978b14f2e9e9 and I can confirm with the test in which it was failing, and it works as expected.

**Comment by r30shah** [comment]
Personal testing has been finished and passed. I am removing WIP

**Review by jdmpapin — APPROVED** [review]

**Comment by jdmpapin** [comment]
Jenkins test sanity all jdk17

**Comment by jdmpapin** [comment]
Whoops, that's not the right Jenkins command. A few things ran (and passed), but I'm starting over to make sure that I'm getting all of the checks that should have run

Jenkins build all

**Comment by jdmpapin** [comment]
@r30shah, I'm not sure whether the slash in https://github.com/eclipse/omr/pull/6329#issuecomment-1029395381 is meant to ask for reviews from both Vijay and myself, or from just one of us. I originally interpreted it as an 'or,' and this change is pretty straightforward, but to me that slash is unclear enough that I figure it's best to ask

**Comment by r30shah** [comment]
@jdmpapin  I originally meant to get atleast Vijay's or your review given the change is made in Value Propagation optimization. Change is trivial enough, if you are OK with the changes, we can merge this one.

**Review by vijaysun-omr — APPROVED** [review]

**Comment by vijaysun-omr** [comment]
I looked at the change now quickly anyway and agree that we can apply it even when vft load is not pattern matched.

**Comment by jdmpapin** [comment]
OK, everything has passed. @r30shah, could you please update the date in the copyright header? I should have checked for that earlier, but if that's the only change then I think we can skip re-testing

**Comment by r30shah** [comment]
Fixed copyright in https://github.com/eclipse/omr/compare/abdec2fc9577133c9b9f9c052b17978b14f2e9e9..dd788de6e3d025f29d715e2d3cf00311751b522c

**Pull request merged**